CIP-????/README.md

```diff
@@ -0,0 +1,471 @@
+- Integers and bytestrings are translated to Plutus Core integers and bytestrings instead of using the `I` and `B` Data constructors.
+- Constructors are translated to Plutus Core `const` terms instead of using the Data `Constr` constructor.
+- Uses of the `Map` and `List` Data constructors will need more specific handling but will be translated in line with their underlying dataype representation in Plutus Core.
+
+## Rationale: how does this CIP achieve its goals?
+
+### Benefits
+
+#### 1. Avoid Decoding the Script Context
+
+This CIP solves problem 1 in the most drastic way possible: it removes it entirely.
+With the proposed ledger interface changes the script context is passed across _already decoded_, so there is zero work for the script to do.
+
+#### 2. Faster Processing of Structured Data
+
+Most Plutus Core programs spend a significant amount of their time operating on datatypes, so we would expect that a performance improvement here would make a significant difference.
+
+Indeed, this seems to be true.
+Our benchmarks[^see-appendix-3] show that this CIP leads to 0-30% real-time speedups depending on the program (the 0% is a primality tester that is mostly doing arithmetic, the 30% is mostly doing datatype manipulation). This is a very significant improvement, and should result in a similar reduction in budget usage.
+
+[^see-appendix-3]: See Appendix 3 for the table of results.
+
+#### 3. Simple Lifting and Unlifting
+
+Today, lifting is complicated, and unlifting is mostly impossible.
+With this CIP lifting becomes very simple, and unlifting becomes not only possible but simple.
+Our prototype implementation contains code-generation to generate lifting and unlifting code in Haskell, but it is straightforward enough that you could easily do it in any language.
+
+There are still some limitations, e.g. you cannot lift or unlift structures that contain functions, but this is not that unusual (similarly, you typically can't serialise such structures).
+
+### Costs
+
+The costs of this proposal are substantial.
+
+Plutus Core today has only 8 kinds of term, this proposal adds 2 more, an increase of 25%.
+Additionally, the new term types are the first term types which have a variable number of children.
+Both of these changes increase implementation complexity throughout the system: everything that processes Plutus Core terms must now handle the new cases, and handle the variable numbers of children.
+
+This change also modestly expends our novelty budget.
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
```

Comment:
I wouldn't be surprised if the Scott-encoded boolean case expression was already faster than the builtin version. As far as I remember, we mostly have the builtin one because we _need_ an eliminator for builtin booleans, not because it's necessarily fast!

Reply:
> I wouldn't be surprised if the Scott-encoded boolean case expression was already faster than the builtin version.

Sure, but that one has to go through the `delay`/`force` hoops, while case-expressions give us enough leeway to bypass that.

> As far as I remember, we mostly have the builtin one because we need an eliminator for builtin booleans, not because it's necessarily fast!

Right... So do we normally use Scott-encoded booleans and only resort to built-in ones when dealing with builtins, which we then/immediately convert to Scott-encoded ones? Can't remember. If that's the case, then we should benefit a little from not Scott-encoding booleans (which I assume we'll get to automatically just by giving up on Scott-encoding in favor of native data types?) and not introducing any `ifThenElse`, but instead inlining the `case` at each call site, so that we can remove `delay`/`force` noise.

Which makes me wonder if there's a lot more pattern matching that we'd better inline during compilation instead of abstracting it out into a matcher.

Or am I saying nonsense?

Reply:
> Sure, but that one has to go through the delay/force hoops, while case-expressions give us enough leeway to bypass that.

Annoyingly, I think we'll still have to put that in. The problem is that the moment you _abstract_ over the case branches (which we do in the compiler), you get strictness again :shrug:  We can probably optimize a lot of it away, though.

> Right... So do we normally use Scott-encoded booleans and only resort to built-in ones when dealing with builtins, which we then/immediately convert to Scott-encoded ones?

Yes.

> which I assume we'll get to automatically just by giving up on Scott-encoding in favor of native data types?

Yes.

> Which makes me wonder if there's a lot more pattern matching that we'd better inline during compilation instead of abstracting it out into a matcher.

We do inline it, but we can't do it until UPLC for annoying type abstraction reasons. We should write some test cases to check that things get as optimized as we like. I have some, but possibly not for this exact case.

Reply:
> As far as I remember, we mostly have the builtin one because we need an eliminator for builtin booleans, not because it's necessarily fast!

It seems like a lost opportunity to not be able to get rid of Builtin types which are sum of products such as booleans. Would it be possible to have a notion of **predefined** datatypes (such as the sop of booleans), and allow builtin functions to return them?

Reply:
I'm not sure what that would look like? Maybe we could allow more of the type language of Typed Plutus Core into the type language of builtin signatures, so they could have anonymous SOP types in there.

We also have to consider how much we're willing to expand the power of builtins. At the moment they cannot construct or deconstruct Plutus Core terms, only pass them around. If we used anonymous SOP types to replace builtin booleans then we'd have to allow _constructing_ Plutus Core terms in builtins. And if we had a builtin function that _takes_ a boolean, we'd have to allow deconstructing Plutus Core terms.

Also, our costing for builtins is based on the assumption that polymorphic arguments can't affect the cost much because the only thing the builtin can do with them is pass them around. I don't know how that changes if it can take a SOP type that it can maybe deconstruct etc.

So basically, lots of complicated details :| Maybe we can do it, but I'm not sure.

Reply:
For the current builtins it seems it would be enough to _return_ a Plutus Core term. Even a Plutus Core _value_ seems enough. All the current (and future) builtins that deconstruct a value could be done with a `case`. The costing of polymorphic functions should stay the same, as long as we require certain reasonable sanity on future built-ins. But yeah, it is a big change...

Reply:
> current builtins

Uh huh, and how about future builtins? Will we allow builtins to case on SOPs or not? We'll have to decide when someone asks. If we don't, we're inconsistent. Making these kind of incrementally-sensible-but-globally-incoherent decisions is how we ended up with the current builtins :D

Reply:
> Will we allow builtins to case on SOPs or not? 

If we want to keep current builtins working, then we'd need casing on a SOP in `IfThenElse` (even though technically `IfThenElse` would be redundant as a builtin).

So currently we convert constants of built-in types to their respective Plutus data types via dedicated converter functions generated by the compiler. I think what Mauro proposes amounts to converting arguments of some predefined types into constants within builtins. Imagine a builtin taking a `Bool` as an example, you'd have to to convert `constr 0` to `False` (or `True`?) somehow. It would be natural to require such a converter alongside the definition of the predefined data type, but then we're at square one -- what's the point of pushing more context into builtins if you end up doing all the same things.

And the denotation of a builtin does take a `Bool` or a `()` or whatever, it doesn't take some arbitrary `constr 0` thing. Imagine a cryptographic function doing something different depending on a `Bool` argument that it has, we _will_ have to convert SoPs to the actual Haskell value.

`Bool` is probably not too complicated though. Something like `[a]` is going to be so much more complex.

So I'm not sure I understand how we'd go about getting rid of data types easily expressible in Plutus and actually do less without forbidding a predefined type to serve as the type of an argument of a builtin.

Reply:
I don't see how this follows. Built-in booleans aren't being removed, no?

Reply:
> I don't see how this follows. Built-in booleans aren't being removed, no?

They are not. And they'll have to stay in order for us to be able to validate old-style scripts regardless. But maybe with SoP we can generate more efficient code by using native Plutus booleans instead of the built-in ones -- that's what we're discussing here.